VmAgent.Core/ProcessOutputLogger.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed under the MIT License.
+
+namespace Microsoft.Azure.Gaming.VmAgent.Core
+{
+    using System;
+    using System.Diagnostics;
+    using global::VmAgent.Core.Dependencies.Interfaces.Exceptions;
+    using Microsoft.Azure.Gaming.VmAgent.Core.Dependencies;
+    using Microsoft.Azure.Gaming.VmAgent.Core.Dependencies.Interfaces;
+
+    public class ProcessOutputLogger: IProcessOutputLogger
+    {
+        private MultiLogger _logger;
+        private IFileWriteWrapper _fileWriteWrapper;
+
+        public ProcessOutputLogger(string logFilePath, MultiLogger logger) 
+            :this (logFilePath, logger, new FileWriteWrapper()) 
+        { 
+        }
+
+        public ProcessOutputLogger(string logFilePath, MultiLogger logger, IFileWriteWrapper fileWriteWrapper)
+        {
+            _logger = logger;
+            _fileWriteWrapper = fileWriteWrapper;
+
+            if (String.IsNullOrWhiteSpace(logFilePath))
+            {
+                throw new ProcessOuputLoggerCreationFailedException($"LogFilePath cannot be null or empty.");
+            }
+
+            try
+            {
+                _fileWriteWrapper.CreateFile(logFilePath);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogException($"Exception was thrown while creating a file. Closing a file {logFilePath}.", ex);
+                Close();
+            }
+        }
+        public void Log(string message)
+        {
+            try
+            {
+                _fileWriteWrapper.Write(message);
+            }
+            catch(Exception ex)
+            {
+                _logger.LogException($"Exception was thrown while writing a log. Closing a file {GetProcessLogFilePath()}.", ex);
+                Close();
+            }
+        }
+
+        public string GetProcessLogFilePath()
+        {
+            return _fileWriteWrapper.GetProcessLogFilePath();
+        }
+
+        public void Close()
+        {
+            _logger.LogInformation($"Closing a Process Log file... Target File Path : {GetProcessLogFilePath()}. ");
+
+            try
+            {
+                if (_fileWriteWrapper.Close())
+                {
+                    _logger.LogInformation($"File is successfully closed.");
+                }
+                else
+                {
+                    _logger.LogInformation($"File is already closed.");
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogException($"Exception was thrown while closing a file..", ex);
+            }
+        }
+
+        public void ErrorOutputHandler(object sendingProcess, DataReceivedEventArgs outLine)
+        {
+            Process p = sendingProcess as Process;
+            if (p != null)
+            {
+                _logger.LogInformation($"ErrorOutputHandler was triggered - ProcessId: {p.Id}");
+            }
+
+            Log(outLine.Data);
+            Close();
```

Comment:
don't close here--this can be called multiple times

Reply:
what could be the case exception is thrown but process is still alive but we don't want to close a file?

Reply:
I was thinking about design that we close file from SessionHostManger but it could be trickier - is there any suggestion to detect process is terminated from event handler?

Reply:
I think you might be misunderstanding what stderr is--
So every process can write to both stdout and stderr. You typically write to stdout using `Console.WriteLine`, which is shorthand for `Console.Out.WriteLine`. In the same way, you can write to stderr using `Console.Error.WriteLine`. Writing to stderr isn't always associated with an Exception being thrown--it's just an output stream a process can write to. It just so happens that by default, uncaught Exceptions get written to stderr by default.
So this `ErrorOutputHandler` method will get called every time the process writes to stderr--not just when the process crashes.

Reply:
I think I missed that point - then this case I should add event handler to `Process.Exited` so they catch errors when Process is terminated. will take a look

Reply:
It might not be necessary to do that--if we're writing stderr to a file, theoretically any uncaught exception in the process will get written to stderr and then written to the file. So when the process exits, the developer will have access to the uncaught exception in the output file.
I think this might even work for the "you entered an invalid start command" thing Yessenia's fix was for--the "that file or process wasn't found" error will get written to stderr and show up in your log file.
Have you tried testing this locally?

Reply:
failed process error is captured before process starts so we don't get errors from event handlers. This is the error when you put a wrong start command. If I understand correctly that fix was for showing the logs when it failed to start because GSDK logs files are not even generated. Correct me if I'm wrong!

``` 
2022-08-23T16:01:10.666-07:00	System.ComponentModel.Win32Exception (2): The system cannot find the file specified.
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at Microsoft.Azure.Gaming.VmAgent.Core.Interfaces.ProcessWrapper.StartWithEventHandler(ProcessStartInfo startInfo, Action`2 StdOutputHandler, Action`2 ErrorOutputHandler, Action`2 ProcessExitedHanlder) in D:\PlayFabGit\MpsAgent\VmAgent.Core\Interfaces\ProcessWrapper.cs:line 48
   at Microsoft.Azure.Gaming.VmAgent.Core.Interfaces.ProcessRunner.CreateAndStart(Int32 instanceNumber, GameResourceDetails gameResourceDetails, ISessionHostManager sessionHostManager) in D:\PlayFabGit\MpsAgent\VmAgent.Core\Interfaces\ProcessRunner.cs:line 72
   ```